Concept: Redirections (#736)
exercism/bash: concepts/redirections/about.md

> # About Redirections

Reviewer: ```suggestion
# About Redirection
```

In the world of the command-line interface (CLI), programs often interact with three standard data streams:

1. **Standard Input** (`stdin`): Where a program receives its input (typically from the keyboard).
1. **Standard Output** (`stdout`): Where a program sends its normal output (typically to the terminal screen).
1. **Standard Error** (`stderr`): Where a program sends error messages (typically to the terminal screen).

Recall that we introduced these in the [Pipelines and Command Lists][pipelines] chapter.

> Shell redirections are powerful mechanisms that allow you to change the default destinations of these streams. Instead of a program reading from the keyboard or writing to the screen, you can redirect input from a file or send output to a file. This provides flexibility and control over how programs interact with data.

Reviewer: ```suggestion
Shell redirection is a powerful mechanisms that allow you to change the default destinations of these streams.
Instead of a program reading from the keyboard or writing to the screen, you can redirect input from a file or send output to a file.
This provides flexibility and control over how programs interact with data.
```

> Shell redirections are powerful mechanisms that allow you to change the default destinations of these streams. Instead of a program reading from the keyboard or writing to the screen, you can redirect input from a file or send output to a file. This provides flexibility and control over how programs interact with data.

Reviewer 2: Is a pipe not considered a form of redirection?

> Shell redirections are powerful mechanisms that allow you to change the default destinations of these streams. Instead of a program reading from the keyboard or writing to the screen, you can redirect input from a file or send output to a file. This provides flexibility and control over how programs interact with data.

Reviewer: Yes, I would say a pipe is another redirection.

Also, this should be single sentence per line, right?

## Key Redirection Operators

Here are the most common redirection operators and what they do:

### `>` (Output Redirection)

* Purpose: Redirects the standard output (stdout) of a command to a file.
* Behavior:
  * If the file doesn't exist, it's created.

>   * If the file exists, its contents are overwritten.

Reviewer 2: Is "overwritten" the same as "truncated"? Does overwritten imply that if the existing file in 10k and we write 1 byte, we only modify one byte?

Reviewer: The effect may be the same, while the work may be less to just overwrite the file.  If we were to truncate, we would have to know information about the content of the file.  To overwrite it we only need to know the name (the file system handles other technical aspects, such as managing fragmentation, etc.).

Reviewer: I think that we can leave it as "overwritten" as it is effectively doing that, to completion.  The technical details are a little deep and involves the media layer.

* Example:
  ```bash
  # Sends the output of 'ls -l' to 'file_listing.txt'
  ls -l > file_listing.txt
  ```

~~~~exercism/caution

> Redirections are performed **before** the command is executed.

Reviewer: ```suggestion
Redirection is performed **before** the command is executed.
```

This does allow for the idea of 1 or more redirection, so the singular seems correct (in this suggestion and the prior).

This means you cannot redirect to the same file you need to read from.

```bash
date > date.txt         # populate the file with the current date
cat date.txt > date.txt # 'date.txt' is now empty!
```

> In the second command of the above example, the shell
> 1. first empties the file named in the redirection, then

Reviewer: ```suggestion
In the second command of the above example, the shell;

1. first empties the file named in the redirection, then
```

> 2. reads zero bytes from the empty file, and writes them to the file.
> ~~~~

Reviewer: ```suggestion
2. reads zero bytes from the empty file, and writes them to the file.

~~~~
```

### `>>` (Append Output Redirection)

* Purpose: Appends the standard output (stdout) of a command to a file.
* Behavior:
  * If the file doesn't exist, it's created.

>   * If the file exists, the output is added to the end of the file.

Reviewer 2: ```suggestion
  * If the file exists, the output is appended to the end of the file.
```

Reviewer: I agree, but "added" is probably generally understood.  Either word works, "added" is more common of a word.

* Example:
  ```bash
  # Adds "Another line" to the end of the file
  echo "Another line" >> file_listing.txt
  ```

### `<` (Input Redirection)

> * Purpose: Redirects the standard input (stdin) of a command from a file.

Reviewer 2: ```suggestion
* Purpose: Redirects the standard input (`stdin`) of a command from a file.
```

* Behavior: The command reads its input from the specified file instead of the keyboard.
* Example:
  ```bash
  # 'sort' reads _from_ 'unsorted_data.txt' and 
  # output is redirected _to_ 'sorted_data.txt'

>   sort < unsorted_data.txt > sorted_data.txt 

Reviewer 2: This is the sort of command that we often scream at people to stop using; sort and many other programs can open files directly. Not that I have a better example at hand.

Reviewer: Yes, similar to abusing cat.  Or abusing ls.

  ```

### `2>` (Error Redirection):

> * Purpose: Redirects the standard error (stderr) of a command to a file.

Reviewer: ```suggestion
* Purpose: Redirect the standard error (stderr) of a command to a file.
```

The purpose is to redirect the standard error of a command to a file.  This could be "of a program to a file" as well, but the point is, it redirects, yes, but the singular should probably be as shown in the  suggestion above.

> * Purpose: Redirects the standard error (stderr) of a command to a file.

Reviewer 2: ```suggestion
* Purpose: Redirects the standard error (`stderr`) of a command to a file.
```

Reviewer: Singluar "Redirect" being my suggestion, but I like the code formatting of the `stderr` here.

* Behavior: Error messages are sent to the specified file instead of the terminal.
* Example:
  ```bash

>   rm non_existent_file 2> error_log.txt # Error message from 'rm' is sent to 'error_log.txt'

Reviewer 2: ```suggestion
  # Error message from 'rm' is sent to 'error_log.txt'
  rm non_existent_file 2> error_log.txt
```

  ```

You can _append_ standard error to a file with `2>>`.

### Additional Redirections

> There are also redirections that can merge both `stdin` and `stderr`. See the [manual][manual] for details. 

Reviewer 2: ```suggestion
There are also redirections that can merge both `stdin` and `stderr`.
See the [manual][manual] for details. 
```

## `/dev/null`

> `/dev/null` is a special file that discards all data written to it.

Reviewer 2: It can also be used as input. It's not only a sink.

Reviewer: I considered commenting on this as well.  Not sure it is important, or if this is fundamentally confusing.  It is just not quite "complete".  I am not sure it needs to be complete.  Unless we decide to show an example where we pipe `/dev/null` to some input.

It is sometimes referred to as "the bitbucket".
It is useful for suppressing output or errors.

```bash
some_command > /dev/null  # Discards stdout
some_command 2> /dev/null # Discards stderr
```

> ## Combining redirections

Reviewer: ```suggestion
## Combining redirection
```

> Multiple redirections can be given at once.

Reviewer: ```suggestion
Multiple redirection can be given at once.
```

I believe this is more correct than using the plural, even here.

Reviewer: It does, but it is a single concept, each redirection can be given multiple times.

```bash
some_command < input.txt > output.txt 2> error.txt
```

> Redirections can be copied.

Reviewer: ```suggestion
Redirection can be copied.
```

```bash
# send stderr to the same destination as stdout
some_command > output.txt 2>&1
```

When you see `2>&1`, read that like "redirect stderr to _whatever stdout is **currently** pointing to_".

> Bash processes redirections strictly from left to right, which means you can do stuff like:

Reviewer: ```suggestion
Bash processes every redirection strictly from left to right, which means you can do things like:
```

```bash
# redirect stdout to 'output1.txt'
# redirect stderr to the same place
# then change stdout to 'output2.txt'
some_command > output1.txt 2>&1 > output2.txt
```

~~~~exercism/advanced
## File Descriptors

> In the context of shell programming and operating systems like Linux/Unix, a file descriptor (FD) is a non-negative integer that serves as an identifier or handle for an open file or I/O resource. Think of it like a label or a reference number that the operating system uses to keep track of open files and data streams.

Reviewer: ```suggestion
In the context of shell programming and operating systems like Linux/Unix, a file descriptor (FD) is a non-negative integer that serves as an identifier or handle for an open file or I/O (Input/Output) resource.
Think of it like a label or a reference number that the operating system uses to keep track of open files and data streams.
```

The numbers you see on the left-hand side of redirections are file descriptors.

The standard I/O streams are given the first 3 file descriptors:

* standard input is FD 0
* standard output is FD 1
* standard error is FD 2

New file descriptors can be set or created with the `exec` command and redirection operators.

```bash
# create FD 3 as a copy of stdout
exec 3>&1
# send stdout to a log file
exec > log.txt

# then redirect both stdout and file descriptor 3
echo "some debug message" >&3  # displayed in the terminal
some_command  2>&1 # stdout and stderr send to 'log.txt'
```

This technique looks quite convoluted, but it can be very useful in situations where your script is being run by some other software package, and standard output is automatically captured by the software.

> The `bats` command works like this.
> ~~~~

Reviewer: ```suggestion
The `bats` command works like this.

~~~~
```

## In Summary

> Why are redirections useful?

Reviewer: ```suggestion
Why is redirection useful?
```

* Saving Output: Store the results of a command for later review or processing.
* Logging: Capture error messages for debugging.
* Automation: Chain commands together, using the output of one as the input of another.
* Filtering: Discard unwanted output.
* Batch Processing: Process data from files instead of typing it manually.
* Scripting: Redirections are essential for writing robust shell scripts.

> 
> 

Reviewer: ```suggestion

```
